### PR TITLE
Added evidence upload and related endpoints to API server and client

### PR DESCRIPTION
**Pull request by Igor8mr**
### Description of the change
Created a new route to the API ('/api/evidence') to organize the evidence endpoints, and added 4 endpoints to it: Get, query, summary and upload. Also added the corresponding commands to the turbinia-client. The first three endpoints are different forms of retrieving information about the TurbiniaEvidence objects in Redis, created in pull request #1330 . The upload endpoint sends a file from the client to the server, so that it can be later processed in a request. This makes it easier to process non-cloud evidence. Unit tests were created to test the new endpoints.

### Applicable issues
Solved issue #1275

### Additional information
This pull request relies on #1330 , so it should be merged after the other is merged.

### Checklist
- [X] All tests were successful.
- [X] Unit tests added.
- [X] Documentation updated.

### CODE REVIEW

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 1]
```diff
@@ -170,6 +187,52 @@ def remove(self, component: MarkdownReportComponent) -> None:
     """
     pass
 
+  def list_to_markdown(self, original_list, level=1):
```
Comment (jleaniz) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:190]: Add docstring here and type annotations.

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 2]
```diff
@@ -170,6 +187,52 @@ def remove(self, component: MarkdownReportComponent) -> None:
...
+      elif isinstance(item, list):
+        report.extend(self.list_to_markdown(item, level + 1))
+      else:
+        if level == 0:
+          report.append(self.heading1(item))
+        else:
+          report.append(self.bullet(item, level))
+    return report
+
+  def dict_to_markdown(
```
Comment (jleaniz) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:206]: Add docstring here and type annotations.

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 3]
```diff
@@ -320,3 +383,100 @@ def __init__(self, request_data: dict):
...
+      report.append(
+          self.bullet(
+              f"Local Path: {evidence_dict.get('local_path', 'null')}", level))
+      report.append(
+          self.bullet(
+              f"Creation Time: {evidence_dict.get('creation_time', 'null')}",
+              level))
+      report.append(
+          self.bullet(
+              f"Last Update: {evidence_dict.get('last_updated', 'null')}",
```
Comment (jleaniz) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:435]: keep consistent with the task attribute name as last_update

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 4]
```diff
@@ -170,6 +187,52 @@ def remove(self, component: MarkdownReportComponent) -> None:
...
+        report.extend(self.list_to_markdown(item, level + 1))
+      else:
+        if level == 0:
+          report.append(self.heading1(item))
+        else:
+          report.append(self.bullet(item, level))
+    return report
+
+  def dict_to_markdown(
+      self, original_dict, level=1, ignore=None, show_null=False,
```
Comment (Igor8mr) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:207]: Check what is ignored and make it more clear

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 5]
```diff
@@ -320,3 +383,100 @@ def __init__(self, request_data: dict):
...
+
+class EvidenceSummaryMarkdownReport(EvidenceMarkdownReport):
+  """Turbinia Evidence Markdown report."""
+
+  def __init__(self, summary: dict | list | int):
+    """Initializes an EvidenceSummaryMarkdownReport object."""
+    super().__init__({})
+    self._summary = summary
+
+  def generate_value_markdown(self, summary, level=1):
```
Comment (jleaniz) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:460]: Add docstring here and type annotations.

File: turbinia/api/cli/turbinia_client/helpers/formatter.py [hunk 6]
```diff
@@ -320,3 +383,100 @@ def __init__(self, request_data: dict):
...
+    report = []
+    for item in summary:
+      self._evidence_data = item
+      report.append(self.generate_markdown(level + 1))
+    return report
+
+  def generate_summary_markdown(self, output='keys'):
```
Comment (jleaniz) [inline turbinia/api/cli/turbinia_client/helpers/formatter.py:467]: Add docstring here and type annotations.

File: turbinia/api/routes/evidence.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+redis_manager = state_manager.RedisStateManager()
+
+EVIDENCE_SUMMARY_ATTRIBUTES = (
+    '_name', 'cloud_only', 'context_dependent', 'copyable', 'creation_time',
+    'description', 'has_child_evidence', 'last_updated', 'local_path',
+    'mount_path', 'parent_evidence', 'request_id', 'resource_id',
+    'resource_tracked', 'save_metadata', 'saved_path', 'saved_path_type',
+    'size', 'source', 'source_path', 'tasks', 'type')
+
+EVIDENCE_QUERY_ATTRIBUTES = EVIDENCE_SUMMARY_ATTRIBUTES + ('tasks',)
```
Comment (jleaniz) [inline turbinia/api/routes/evidence.py:43]: tasks is already in EVIDENCE_SUMMARY_ATTRIBUTES

File: turbinia/api/routes/evidence.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        processing. The maximum size of the file is 10 GB. 
+  
+  Returns:
+    List of uploaded evidences or warning messages if any.
+  """
+  evidences = []
+  for file in files:
+    warning_message = None
+    try:
+      file_path = await get_file_path(file.filename, ticket_id)
```
Comment (jleaniz) [inline turbinia/api/routes/evidence.py:236]: file_path would not be accessible below if an OSError exception is thrown. I would either initialize file_path to None or consolidate both try statements in one.

File: turbinia/config/__init__.py [hunk 9]
```diff
@@ -68,7 +68,10 @@
     'API_SERVER_PORT',
     'API_ALLOWED_ORIGINS',
     'API_AUTHENTICATION_ENABLED',
-    'WEBUI_PATH',
+    'CHUNK_SIZE',
```
Comment (jleaniz) [inline turbinia/config/__init__.py:71]: prefix the config variable name with API_

API_UPLOAD_CHUNK_SIZE
API_EVIDENCE_UPLOAD_DIR 

etc

**Review by jleaniz — CHANGES_REQUESTED** [review]
Thanks for the PR, i left some comments to review.

**Comment by jleaniz** [comment]
@Igor8mr you should be able to merge the master branch now into this PR to include the required changes from #1330, pending merge conflict resolutions.

**Review by jleaniz — APPROVED** [review]
LGTM

**Pull request merged**